Prioritize controllers with an intents array for the same state (#241)
VoxaAI/voxa: docs/voxa-app.rst

    });
    voxaApp.onState('ProcessUserRequest', (voxaEvent) => {
      // Some other code
      return { tell: 'GoodbyeResponse', to: 'die' };
    });

  If the state machine goes to the `ProcessUserRequest`, the code that will run always will be the first one, so the user will always hear the `ThankYouResponse`.

  The only scenario where this is overrided is when you have two states with the same name and one of them has one or more intents defined to handle. If the user triggers the intent inside the defined intents, Voxa will give priority to the state with intents. For example, take this code:

  .. code-block:: javascript

    voxaApp.onState('ProcessUserRequest', (voxaEvent) => {
      // Some code
      return { tell: 'ThankYouResponse', to: 'die' };
    });
    voxaApp.onState('ProcessUserRequest', (voxaEvent) => {
      // Some other code
      return { tell: 'GoodbyeResponse', to: 'die' };
    }, "YesIntent");

  If the user triggers a `YesIntent`, and the state machine goes to the `ProcessUserRequest` state, the user will listen the `GoodbyeResponse`, it doesn't matter if the code is above or below the other state.

> 

Reviewer: You can set multiple **controllers** for a single **state**. Also, might worth showing an example with more than just 2 controllers for a state, something like

```js

voxaApp.onState("agreed?", {
  to: "PurchaseAccepted"
}, "YesIntent");

voxaApp.onState("agreed?", {
  to: "TransactionCancelled"
}, ["NoIntent", "CancelIntent"]);

voxaApp.onState("agreed?", {
  to: "agreed?",
  reply: "Help.ArticleExplanation",
  flow: "yield"
}, "HelpIntent");

voxaApp.onState("agreed?", {
  to: "agreed?",
  reply: "UnknownInput",
  flow: "yield"
})

Author: Made some changes. Let me know what you think.

      return { tell: 'LaunchIntent.OpenResponse', to: 'die' };
    });

>   Also you can use a shorthand version to define a state. This is very useful when having a state that only returns a :ref:`transition <transition>`

Reviewer: ```suggestion
  Also you can use a shorthand version to define a controller. This is very useful when having a controller that only returns a :ref:`transition <transition>`
```

  .. code-block:: javascript

    voxaApp.onState('launch',
      {
        flow: 'yield'
        reply: 'LaunchIntent.OpenResponse',
        to: 'nextState'
      }
    );

>   You can also set the intent that the state will handle. If set, any other triggered intent will not enter into the state.

Reviewer: ```suggestion
  You can also set the intent that the controller will handle. If set, any other triggered intent will not enter into the controller.
```

  .. code-block:: javascript

    voxaApp.onState("agreed?", {
      to: "PurchaseAccepted"
    }, "YesIntent");

    voxaApp.onState("agreed?", {
      to: "TransactionCancelled"
    }, ["NoIntent", "CancelIntent"]);

    voxaApp.onState("agreed?", {
      to: "agreed?",
      reply: "Help.ArticleExplanation",
      flow: "yield"
    }, "HelpIntent");

    voxaApp.onState("agreed?", {
      to: "agreed?",
      reply: "UnknownInput",
      flow: "yield"
    });

>   **The order on how you structure your states matter in Voxa**

Reviewer: ```suggestion
  **The order on how you declare your controllers matter in Voxa**
```

  You can set multiple :ref:`controllers <controllers>` for a single state, so how do you know which code will be executed? The first one that Voxa finds. Take this example:

  .. code-block:: javascript

    voxaApp.onState('ProcessUserRequest', (voxaEvent) => {
      // Some code
      return { tell: 'ThankYouResponse', to: 'die' };
    });
    voxaApp.onState('ProcessUserRequest', (voxaEvent) => {
      // Some other code
      return { tell: 'GoodbyeResponse', to: 'die' };
    });

>   If the state machine goes to the `ProcessUserRequest`, the code that will run always will be the first one, so the user will always hear the `ThankYouResponse`.

Reviewer 2: *the code running will always be the first one

>   If the state machine goes to the `ProcessUserRequest`, the code that will run always will be the first one, so the user will always hear the `ThankYouResponse`.

Reviewer 2: ```suggestion
  If the state machine goes to the `ProcessUserRequest`, the code running will always be the first one, so the user will always hear the `ThankYouResponse`.
```

>   The only scenario where this is overrided is when you have two controllers for the same state, and one of them has one or more intents defined to handle. If the user triggers the intent that's inside the defined intents Voxa will give priority to the state with intents. For example, take this code:

Reviewer: ```suggestion
  The only scenario where this is overrided is when you have more than one for the same state, and one of them has one or more intents defined. If the user triggers the intent that's inside the list of intents of one controller Voxa will give it priority. For example, take this code:
```

>   The only scenario where this is overrided is when you have more than one for the same state, and one of them has one or more intents defined. If the user triggers the intent that's inside the list of intents of one controller Voxa will give it priority. For example, take this code:

Reviewer 2: * overwritten instead of overrided
* more than one .......handler?
* list of one-controller intents, Voxa

>   The only scenario where this is overrided is when you have more than one for the same state, and one of them has one or more intents defined. If the user triggers the intent that's inside the list of intents of one controller Voxa will give it priority. For example, take this code:

Reviewer 2: ```suggestion
  The only scenario where this is overwritten is when you have more than one handler for the same state, and one of them has one or more intents defined. If the user triggers the intent that's inside the list of one-controller intents, Voxa will give it priority. For example, take this code:
```

  .. code-block:: javascript

    voxaApp.onState("agreed?", {
      to: "PurchaseAccepted"
    }, "YesIntent");

    voxaApp.onState("agreed?", {
      to: "agreed?",
      reply: "UnknownInput",
      flow: "yield"
    });

    voxaApp.onState("agreed?", {
      to: "TransactionCancelled"
    }, ["NoIntent", "CancelIntent"]);

>   If the user triggers the `NoIntent`, and the state machine goes to the `agreed?` state, the user will listen the `TransactionCancelled` response, it doesn't matter if the controller is placed above or below a controller without defined intents, the priority will go to the controller with the defined intent.

Reviewer 2: * listen to

>   If the user triggers the `NoIntent`, and the state machine goes to the `agreed?` state, the user will listen the `TransactionCancelled` response, it doesn't matter if the controller is placed above or below a controller without defined intents, the priority will go to the controller with the defined intent.

Reviewer 3: ```suggestion
  If the user triggers the `NoIntent`, and the state machine goes to the `agreed?` state, the user will listen to the `TransactionCancelled` response, it doesn't matter if the controller is placed above or below a controller without defined intents, the priority will go to the controller with the defined intent.
```
```suggestion
  If the user triggers the `NoIntent`, and the state machine goes to the `agreed?` state, the user will listen the `TransactionCancelled` response, it doesn't matter if the controller is placed above or below a controller without defined intents, the priority will go to the controller with the defined intent.
```

>   If the user triggers the `NoIntent`, and the state machine goes to the `agreed?` state, the user will listen the `TransactionCancelled` response, it doesn't matter if the controller is placed above or below a controller without defined intents, the priority will go to the controller with the defined intent.

Reviewer 2: ```suggestion
  If the user triggers the `NoIntent`, and the state machine goes to the `agreed?` state, the user will listen to the `TransactionCancelled` response, it doesn't matter if the controller is placed above or below a controller without defined intents, the priority will go to the controller with the defined intent.
```
